### PR TITLE
UI: Installing, validating and testing

### PR DESCRIPTION
**Pull request by SiAdcock**
## What does this change?

- [x] Install `ui` assets on TeamCity
- [x] Validate `ui/` directory on pre-commit
- [x] Node-ify pre-push (this was previously a bash script)
- [x] Validate `ui/` directory on pre-push

## What is the value of this and can you measure success?

git-hooks and the test runner will now validate and test the `ui/` directory

It would be nice if the test runner only ran validate and test tasks against files that have changed locally (or if there is no corresponding remote branch, files that have changed in the current versus master). However I'd like to keep this PR as small as possible for now.

### CODE REVIEW

File: git-hooks/pre-commit [hunk 1]
```diff
@@ -12,8 +12,13 @@ execa
...
-            .filter(file => file.startsWith('static') && file.endsWith('.js'))
+            .filter(
+                file =>
+                    (file.startsWith('static') || file.startsWith('ui')) &&
+                    (file.endsWith('.js') || file.endsWith('.jsx'))
+            )
             .filter(file => fs.existsSync(file));
+
         if (jsFiles.length) {
             return execa('eslint', ['--fix', ...jsFiles], {
```
Comment (sndrs) [inline git-hooks/pre-commit]: this will use the main installation of eslint, which `ui` doesn't use currently

**Comment by SiAdcock** [comment]
@sndrs I believe this is now ready for a proper review, when you have time. The PR is becoming quite large, so I've de-scoped the plan to only test/validate changed files on `pre-push`. Perhaps this is an optimisation we can add in a future change.

File: git-hooks/pre-commit [hunk 2]
```diff
@@ -6,35 +6,65 @@ const fs = require('fs');
...
-                        stdio: 'inherit',
-                    })
-                )
-                .then(() => staged);
-        }
-        return staged;
-    })
-    // make sure yarn.lock is commited
+    .then(staged =>
+        fixStagedIn(staged, 'static').then(() => fixStagedIn(staged, 'ui'))
```
Comment (SiAdcock) [inline git-hooks/pre-commit]: Simply code DRYness. Is it difficult to follow?

Comment (sndrs) [inline git-hooks/pre-commit]: don't we just need the list of files, then pass that whole thing to eslint to manage? don't need 2 tasks to do it?

File: git-hooks/pre-push [hunk 3]
```diff
@@ -1,50 +1,88 @@
...
+                (prev, curr) => prev.then(curr),
+                Promise.resolve()
+            );
+        });
+    }
+
+    return Promise.resolve();
+};
+
+getCurrentBranchName().then(branch =>
```
Comment (sndrs) [inline git-hooks/pre-push]: `Promise.all([getCurrentBranchName(), getRemoteBranches()])`?

**Comment by sndrs** [comment]
do you think we should remove the task runner bit from push? effectively running the same task as commit but on the branch diff?

if we did that, do we need the commit hooks?

or maybe create a more encompassing push task that checks on the diff, offers to fix and commit etc if necessary, bails if it cannot fix?

**Comment by SiAdcock** [comment]
Superseded by #17647

**Pull request closed without merging**